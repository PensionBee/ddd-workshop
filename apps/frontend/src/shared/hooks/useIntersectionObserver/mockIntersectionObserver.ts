export const mockDisconnect = jest.fn();
export const mockObserve = jest.fn();
export const mockUnobserve = jest.fn();

export class MockIntersectionObserver {
  readonly root: Element | null;
  readonly rootMargin: string;
  readonly thresholds: ReadonlyArray<number>;
  constructor() {
    this.root = null;
    this.rootMargin = "";
    this.thresholds = [];
  }
  disconnect() {
    mockDisconnect();
  }
  observe() {
    mockObserve();
  }
  takeRecords(): IntersectionObserverEntry[] {
    return [];
  }
  unobserve() {
    mockUnobserve();
  }
}
