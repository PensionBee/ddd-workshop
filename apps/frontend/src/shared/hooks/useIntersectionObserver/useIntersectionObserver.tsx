import type { RefObject } from "react";
import { useEffect, useState, useRef } from "react";

export const useIntersectionObserver = (
  ref: RefObject<Element>,
  freeze?: boolean
) => {
  const [isOnScreen, setIsOnScreen] = useState(false);
  const observerRef = useRef<null | IntersectionObserver>(null);

  const cleanup = () => {
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    observerRef && observerRef.current && observerRef.current.disconnect();
  };

  useEffect(() => {
    const hasIOSupport = Boolean(window.IntersectionObserver);

    if (!hasIOSupport) return setIsOnScreen(true);

    observerRef.current = new IntersectionObserver(([entry]) => {
      setIsOnScreen(entry.isIntersecting);
      if (!freeze) return;
      // After observation, clear observer if freeze is true
      return cleanup();
    });
  }, [freeze]);

  useEffect(() => {
    ref.current &&
      observerRef.current &&
      observerRef.current.observe(ref.current);

    return () => cleanup();
  }, [ref]);

  return isOnScreen;
};
