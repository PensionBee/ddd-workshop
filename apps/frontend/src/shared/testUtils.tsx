import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
