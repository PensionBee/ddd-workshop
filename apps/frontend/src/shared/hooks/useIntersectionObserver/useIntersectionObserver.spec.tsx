import { render } from "@shared/testUtils";
import { useRef } from "react";
import {
  MockIntersectionObserver,
  mockObserve,
} from "./mockIntersectionObserver";
import { useIntersectionObserver } from "./useIntersectionObserver";

describe("useIntersectionObserver", () => {
  describe("Default", () => {
    it("should call observe", () => {
      window.IntersectionObserver = MockIntersectionObserver;
      const Component = () => {
        const div = useRef(null);
        const isVisible = useIntersectionObserver(div);
        return <div ref={div}>{isVisible ? "true" : "false"}</div>;
      };
      const { getByText, asFragment } = render(<Component />);

      // expect(mockObserve.mock.calls.length).toBe(1);
      expect(getByText("false")).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });

    it("should default visible with no intersection observer", () => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (window as any).IntersectionObserver = undefined;
      const Component = () => {
        const div = useRef(null);
        const isVisible = useIntersectionObserver(div);
        return <div ref={div}>{isVisible ? "true" : "false"}</div>;
      };
      const { getByText, asFragment } = render(<Component />);

      // expect(mockObserve.mock.calls.length).toBe(0);
      expect(getByText("true")).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
