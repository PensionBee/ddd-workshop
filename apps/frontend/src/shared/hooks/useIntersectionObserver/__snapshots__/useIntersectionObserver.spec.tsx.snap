// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useIntersectionObserver Default should call observe 1`] = `
<DocumentFragment>
  <div>
    false
  </div>
</DocumentFragment>
`;

exports[`useIntersectionObserver Default should default visible with no intersection observer 1`] = `
<DocumentFragment>
  <div>
    true
  </div>
</DocumentFragment>
`;
